Simplify precision-specification in tests.

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37181 Simplify precision-specification in tests.**

Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

Differential Revision: [D21728402](https://our.internmc.facebook.com/intern/diff/D21728402)

nairbv committed:
Simplify precision-specification in tests.

Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

nairbv commented:
@ezyang @zou3519 @gchanan There's still (at least one) test failure in this PR (`TestTensorDeviceOpsCUDA.test_qr_big_cuda_float32` on windows). Before thinking about what to do about that case though, I'd like to check for feedback on the general approach here.

Personally I think this is simpler, in most cases tightens the tolerances, and makes the low-tolerance ops standout a bit more. It also reduces flexibility a bit though, and in at least some cases loosens tolerance slightly.

zou3519 commented:
What is your plan of record of how we should override precision?

Right now from the diff it seems like if we want to override a test in `tensor_op_tests`, then we should "put the operator into the `very_low_precision_half_cuda_tests` list". This seems bad for readability to me: we can't tell "from a glance" at the `torch.remainder` tests what the precision overrides are; we have to manually scroll down and see if `remainder` is in "very_low_precision_half_cuda_tests".

Why can't we leave the overrides in `tensor_op_tests`, but make it optional to write them?

nairbv commented:
> What is your plan of record of how we should override precision?

I'd generally rather avoid overriding precision -- I think we do it too much and in too many ways. My thinking is that results should almost always fall within some reasonable default tolerance for a given dtype. On master we have operators being tested at floating tolerances as high as `1.0`, and I think we should strongly discourage that, in part by making it difficult. A rare special test could be written for an outlier operator that really needs special handling, but it seems problematic to specify whatever-we-want on every test case.

> Right now from the diff it seems like if we want to override a test in `tensor_op_tests`, then we should "put the operator into the `very_low_precision_half_cuda_tests` list".

This is only for half-cuda, and even here I'd kind of prefer to widen the default tolerance on half until this isn't necessary, though I'm not sure how others feel about that. i.e. the fact that so many tests don't meet the tolerance we set for half makes me think 5e-2 could be a better default than 1e-3, allowing us to remove the special case. Tests for all other dtypes (besides the one test failing on windows) fall within the defaults, and I'd prefer to keep it that way.

> This seems bad for readability to me: we can't tell "from a glance" at the `torch.remainder` tests what the precision overrides are;

Some things I find unreadable about the current state: 
* I can't easily tell which operators deviate from default tolerances.
* If the defaults are changed, values specified here still override those defaults.
* It's a very large table with a lot of parens. Minimizing the special cases just seems easier on the eyes, and modification seems error-prone.
* Many of these are specifying tolerances for dtypes that aren't even being tested.

> we have to manually scroll down and see if `remainder` is in "very_low_precision_half_cuda_tests".

True, there are some trade-offs here. Personally I'm valuing "find the few outliers if any" over "see details in every case," though another option is to just widen for all half-cuda tests. We may also find ways to make some of these operators more accurate.

> Why can't we leave the overrides in `tensor_op_tests`, but make it optional to write them?

The table is positional. If we want them to be optional they either need to be specified as None or be the last parameters in the list. There are already a lot of optional parameters that, if needed, force specifying all prior ones. That's why so many already explicitly specify tolerances that are the defaults.

zou3519 commented:
> I'd generally rather avoid overriding precision -- I think we do it too much and in too many ways. 

What do you think are the valid cases where someone should be allowed to override precision? half operations seems like one because sometimes it is difficult to write code to get the best precision. I'm trying to understand if we can make the blanket statement "you should never override precision" or not. 

> This is only for half-cuda, and even here I'd kind of prefer to widen the default tolerance on half until this isn't necessary, though I'm not sure how others feel about that. i.e. the fact that so many tests don't meet the tolerance we set for half makes me think 5e-2 could be a better default than 1e-3, allowing us to remove the special case.

half precision is advertised as being correct to a little more than 3 decimal digits (https://en.wikipedia.org/wiki/Half-precision_floating-point_format). Reducing the tolerance is probably a bad idea because it is good that we know some of our code is correct to the 3 decimal places. On the other hand, in a lot of nn models the precision doesn't matter...

ezyang commented:
I haven't thought about the precision issue enough yet, but...

> The table is positional. If we want them to be optional they either need to be specified as None or be the last parameters in the list. There are already a lot of optional parameters that, if needed, force specifying all prior ones. That's why so many already explicitly specify tolerances that are the defaults.

There's a simple way to solve this problem: stop specifying the table as tuples. A nice pattern that I like is to make table rows function calls to some short function name, like `r(1, 2, foo=3)` and now you can mix positional and keyword arguments.

ezyang commented:
I'm with Richard that if we *do* have to override the precision for a test, I'd prefer we do it precisely where the test is. However, in general, I think you, @nairbv, are going to have to make the call about what exactly to do about precisions in general. Most of what you've written here seems reasonable to me; I guess there is only one more thing I would add, which is that our test precision specification has never been *prescriptive*; it's been entirely descriptive (people tweaking the precisions until the tests passed). Because of this, I am a bit wary about relaxing precisions, because whenever you relax the precision, you are erasing the history of a test in a way that is not easily recoverable. If we're throwing out information about tests, we should have a better reason than just "it looks better if we don't have to override as many".

ezyang reviewed (approved):
However, I think this PR is an improvement, so I am going to permit it.

nairbv commented:
> However, I think this PR is an improvement, so I am going to permit it.

That's my initial thinking, that a number of tests getting tightened from precision of 1.0 to 5e-2 outweights a few tests get widened from, say, 1e-3 to 5e-2, but I also think you're right that finding a better way to express the deviations seems reasonable, and would solve for the case of the one-off windows test that's currently failing, so I can add that.

> half precision is advertised as being correct to a little more than 3 decimal digits

I think this in particular is an argument that the half-precision I set initially is too tight. e.g. float holds a bit over 8 decimal digits, so we set an atol of 1e-5. The result is going to lose precision from being operated on. I had set a default rtol of 0.1% and an atol of +/- 1e-5  for half. Intuitively it seems like a value that can only even accurately store 3 decimal digits is only confidently going to generate results within an rtol of, say, 1% of the target value, though having a very low atol allows validating results close to zero.

Regardless of the details of each individual test, I think we should be careful about setting defaults to something that will make sense and would be reasonable for the majority of tests on the given dtype.

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

ezyang commented:
I doubt your whitespacing will pass lint

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

nairbv commented on torch/testing/_internal/common_utils.py:
```diff
@@ -885,7 +885,8 @@ def safeCoalesce(self, t):
     dtype_precisions = {
-        torch.float16    : (0.001, 1e-5),
+
+        torch.float16    : (0.005, 1e-4),
```
@zou3519 FYI I made a small adjustment to tolerances here in this PR (post-approval) based on the idea that a half type holds about 3 decimal digits. This rtol might still be a bit too tight of a default for many half cases, but we can always relax it more later if we find ourselves needing to tweak too often.
I think it makes sense for bfloat to have a tighter default atol than half since it has higher dynamic range (the same exponent bits as float32).

zou3519 replied:
Since half type holds more than 3 decimal places, 0.001 seems like a good number for `rtol`. Were there specific cases where a test needs more than 0.001 for `rtol`?

Note that it's not that "a half type holds about 3 decimal digits", to be more exact, a half type holds "more than 3 decimal digits": https://en.wikipedia.org/wiki/Half-precision_floating-point_format

zou3519 replied:
If there's anything that needed something less tight than `(0.001, 1e-5)`, then using the `half_tol `override mechanism would preserve the old behavior without relaxing all of the tolerances

nairbv replied:
> Since half type holds more than 3 decimal places

I don't think that's correct, "safely holds" doesn't mean it will retain that precision after an operation. We don't check tolerances of 1e-7 on float, nor 1e-15 on double, because so many of our operators are non-trivial matrix operations and we don't want to have to manually set tolerances on every test. "Safely holds 3" is why I wanted to widen it and why even 0.005 seems pretty tight.

E.g. testing a simple sum operation on moderately sized half tensor already brings us to a tolerance of ~0.004:
```
r=torch.arange(0, 99, dtype=torch.float)
r.sum().item()
Out[63]: 4851.0
r.to(torch.half).sum().item() / r.sum().item() - 1
Out[64]: -0.003916718202432445
```
Requiring tests to special-case the tolerance on tests this simple feels like it creates excessive overhead and complication across the board.

> using the half_tol override mechanism

My motivation for loosening it wasn't specifically for these tests, but because I think it's a more reasonable default. It would be easy to adapt these tests to any default tolerance, I just don't feel like a tenth of a percent of tolerance makes sense when checking results of operations on a dtype that doesn't precisely represent `3333`.

zou3519 replied:
`torch.sum` on a CPU half tensor is not the best example because we don't expect users to use CPU half. We don't do dtype accumulation on CPU half tensors for torch.sum, but we do accumulate into float for CUDA half tensors, so the precision on torch.sum for CPU half tensors is "bad".

The reason why I am wary about relaxing precisions is it's possible that in the future we regress in precision (an operation that was able to support 0.001 rtol before doesn't anymore) and this causes some pre-existing model to not converge anymore. It would be difficult to bisect to such a commit.

That's the conservative side in me speaking and I am not sure how realistic that situation is. But I've spent a good chunk of time in the past debugging why models don't converge anymore on new releases of PyTorch. My personal opinion is that I am not comfortable with relaxing pre-existing half precision testing across the board.

nairbv replied:
> relaxing pre-existing half precision

I wouldn't exactly think of it as pre-existing. I do want to choose _initial_ defaults that make sense for PyTorch broadly. This is a new(ish) parameter, and when I added it I felt like it was easier to be conservative and relax later than to start loose and tighten later. Many (most?) half tolerances are currently being set manually through atol, which feels unsustainable to me.

If we can't pick a tolerance that fits a super-majority of operators then I suspect we're doing something wrong.

> so the precision on torch.sum for CPU half tensors is "bad".

That's interesting, I had tried it on a non-cuda host. If we can reasonably expect most overriding operators fit a tighter tolerance than they do now, maybe there should be a project to fix "bad" operators? I suspect that some of the overrides even outside this looser 0.005 tolerance are also "bad" and should be fixed (and was hoping we'd revisit some of these once this lands and they stand out better).

I think I'll feel more comfortable setting a tighter tolerance if we can feel comfortable asserting that operators outside that tolerance are exceptional/bad in some way. That seems like a hard argument to make when a substantial portion of our tests are already exceptions.

A standard for tolerance should enforce "correctness" under some reasonable default definition of correctness. When most operators are tested merely for whatever worked at the time it was written, then we're only testing for regressions and not that it's not "bad."

A common rarely-overridden default provides a stronger incentive not to land "bad" operators in the first place.

zou3519 replied:
I see, I didn't realize how recent the rtol default was. You've also changed the atol default though (from 1e-5 to 1e-4), does that relax any pre-existing tolerances, and if so, by how much? If it turns out that we haven't relaxed it by very much then my concern about regressions becomes moot.

nairbv replied:
The original default self.precision was set to 1e-5 and applied as atol, but I don't think it was successfully used much with half-precision tests. The absolute difference between 1/3 stored in a half and 1/3 as a double is greater than 1e-5, but less than 1e-4, even without performing any further operations.

I am somewhat swayed by the argument that cpu-sum is "bad," if special-handling code for `sum` and similar operators is highlighting a real problem.

Re-running this batch of tests with default rtol set back to 0.001, the test failures I see that we would instead need to explicitly call out with an override are TestTensorDeviceOpsCUDA.:
test_addbmm_cuda_float16
test_addbmm_inplace__cuda_float16
test_addbmm_scalar_cuda_float16
test_addbmm_inplace_scalar_cuda_float16
test_addcdiv_inplace_scalar_cuda_float16
test_addcdiv_scalar_cuda_float16
test_cumsum_cuda_float16
test_cumsum_neg_dim_cuda_float16
test_prod_cuda_float16

If we suspect addbmm, addcdiv, cumsum, and prod have differences due to normal/unimportant/minor precision variations then I'd prefer to continue with 0.005. If we think these are issues like cpu-sum that we can consider "bad" that we could/should "fix," then I'm pretty amenable to lowering back to 0.001. What do you think?

zou3519 replied:
@nairbv my concern is only about any potential relaxing of the half precision (i.e., if we were testing something with tighter tolerances before than we are after this PR). You've addressed that concern so I don't have more concerns.

Regarding the default, I don't know "what a good default" should be. Looking at the list of ops above, cumsum doesn't actually accumulate into float (unlike a lot of our other ops, but it's been like that for a few years. Maybe some of the folks on https://github.com/pytorch/pytorch/pull/36458 would know more). `prod` accumulates into `float` on half types, so I'm not sure why it isn't passing rtol=0.001 (https://github.com/pytorch/pytorch/blob/0ff1aa90589ed113ee127dea484f77f3a014efc4/aten/src/ATen/native/cuda/ReduceSumProdKernel.cu#L46). `addbmm` should call straight into the CUDA API so we should at least be able to say that that one is "implemented correctly"

nairbv commented on test/test_torch.py:
```diff
@@ -17837,407 +17838,435 @@ def inner(self, device, dtype):
-        1e-1, 1e-1, 1e-4, _float_types2),
-    ('addbmm', 'scalar', _small_2d, lambda t, d: [_number(0.4, 2, t), _small_3d(t, d), _small_3d(t, d)],
-        1e-1, 1e-1, 1e-4, _float_types2, _cpu_types, True,
+    _tc('add', '', _small_3d, lambda t, d: [_number(3.14, 3, t)]),
```
@ezyang / @zou3519 I converted this to function calls (post-approval) to keep overrides with the test definitions and allow fine-grained adjustment per-test. Let me know if any objections before I land.
There are frequent new conflicts, so if you see minor non-critical issues I can always add a follow-up PR too.

zou3519 replied:
No objections from me for the function calls, they're pretty nice

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -17838,407 +17839,435 @@ def inner(self, device, dtype):
+        self.dtypes = _types if dtypes is None else dtypes
+        self.cpu_dtypes = [] if cpu_dtypes is None else cpu_dtypes
+        self.has_inplace = has_inplace
+        self.decorators = [] if decorators is None else decorators
```
nit: The defaults for `ctor`, `dtypes`, `cpu_dtypes`, `decorators` should appear in the `__init__` constructor.

If the linter complains, then that's because any defaults cannot be a list. Although the defaults can't be lists, they can still be (and should be) tuples instead.

I'm fine with handling this in a follow-up if you're trying to avoid merge conflicts

nairbv replied:
You're correct that I moved this due to lint errors. I see there's an easy way to convert a list to a tuple, didn't realize that was an option. I'd probably move that to a followup PR though.

zou3519 commented on test/test_torch.py:
```diff
@@ -17838,407 +17839,435 @@ def inner(self, device, dtype):
     # Constructs decorator list and applies decorators
-    if decorators is None:
-        decorators = [dtypes(*dtype_list)]
+    if testMeta.decorators is None:
```
afaict, `testMeta.decorators` is never None. It's always a list (and can be empty)

zou3519 commented on test/test_torch.py:
```diff
@@ -18248,32 +18277,13 @@ def fn(self, device, dtype):
+    def caller(cls, testMeta):
+        generate_test_function(cls, testMeta)
+        if testMeta.has_inplace:
+            generate_test_function(cls, testMeta.inplace)
```
nit: `testMeta.inplace_variant`, since you've made a helper function?

zou3519 commented on test/test_torch.py:
```diff
@@ -17838,407 +17839,435 @@ def inner(self, device, dtype):
-                           dtype_list,
-                           dtype_cpu_list,
-                           decorators):
+def generate_test_function(cls, testMeta):
```
nit: In python variable names are usually snake_case

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

Differential Revision: [D21728402](https://our.internmc.facebook.com/intern/diff/D21728402)

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

Differential Revision: [D21728402](https://our.internmc.facebook.com/intern/diff/D21728402)

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

Differential Revision: [D21728402](https://our.internmc.facebook.com/intern/diff/D21728402)

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

Differential Revision: [D21728402](https://our.internmc.facebook.com/intern/diff/D21728402)

[ghstack-poisoned]

nairbv committed:
Update on "Simplify precision-specification in tests."


Now that assertEquals considers dtypes in determining tolerance, most
tests don't need explicitly set precision.

Those that do are a few half precision tests on cuda. In this PR, those
are broken out to be handled explicitly, though we may also want to
consider further loosening the tolerance on half-precision.

Differential Revision: [D21728402](https://our.internmc.facebook.com/intern/diff/D21728402)

[ghstack-poisoned]

jeffdaily commented:
@ezyang @nairbv @rohithkrn This PR breaks bfloat16 tests for ROCm CI.

jeffdaily commented:
CC @sunway513

ngimel commented:
I'm going to revert this PR because it breaks ROCm, and we don't want to shadow other potential breakages.
